flagscale/train/arguments.py

```diff
@@ -0,0 +1,228 @@
+
+
+import torch
+import types
+import ast
+import itertools
+from datetime import timedelta
+
+import torch
+
+from flagscale.train.hetero.parallel_context import RankMapper
+
+
+class FSTrainArguments:
+    """Extend the Megatron arguments with FlagScale specific arguments.
+    """
+    
+    def __init__(self, args, rank_mapper=None):
+        self.args = args
+        self._rank_mapper = rank_mapper
+
+    def __getattr__(self, name):
+        if name == "rank_mapper":
+            return self._rank_mapper
+        return getattr(self.args, name)
+
+    def __str__(self):
+        return str(self.args)
+
+    def __repr__(self):
+        return repr(self.args)
+    
+    def _initialize_distributed(self):
+        """Initialize torch.distributed and core model parallel."""
+        args = self.args
+
+        device_count = torch.cuda.device_count()
+        if torch.distributed.is_initialized():
+
+            if args.rank == 0:
+                print(
+                    "torch distributed is already initialized, "
+                    "skipping initialization ...",
+                    flush=True,
+                )
+            args.rank = torch.distributed.get_rank()
+            args.world_size = torch.distributed.get_world_size()
+
+        else:
+
+            if args.rank == 0:
+                print("> initializing torch distributed ...", flush=True)
+            # Manually set the device ids.
+            if device_count > 0:
+                torch.cuda.set_device(args.local_rank)
+                device_id = torch.device(f'cuda:{args.local_rank}')
+            else:
+                device_id = None
+
+            # Call the init process
+            init_process_group_kwargs = {
+                'backend' : args.distributed_backend,
+                'world_size': args.world_size,
+                'rank': args.rank,
+                'timeout': timedelta(minutes=args.distributed_timeout_minutes),
+            }
+            torch.distributed.init_process_group(**init_process_group_kwargs)
+    
+    
+    def _build_rank_mapper(self):
+        self._initialize_distributed()
+        self._rank_mapper = RankMapper(self.args)
+        return self._rank_mapper
+
+    @classmethod
+    def from_dict(cls, args):
+        return cls(args)
+
+    def pre_validate_args(self):
```

Comment:
The arguments extended by flagscale were split into two parts for now. However, in the future, the arguments may have more than two parts.

Reply:
To decrease the code change in the `Megatron-LM` repo, limit modifications to the `validate_args` function in the original argument file when adding new parameters.